File: lua/anagrams.lua
function generatePermutations(a, n)
  if n == 0 then
    print(utf8.char(table.unpack(a)))
  else
    for i = 1, n-1 do
      generatePermutations(a, n-1)
      local j = n % 2 == 0 and i or 1
      a[j], a[n] = a[n], a[j]
    end
    generatePermutations(a, n-1)
  end
end

if #arg ~= 1 then
  io.stderr:write('Exactly one argument required\n')
  os.exit(1)
end
word = {utf8.codepoint(arg[1], 1, utf8.len(arg[1]))}
generatePermutations(word, #word)

File: dart/anagrams.dart
import 'dart:io';

void main(List<String> args) {
  if (args.length != 1) {
    stderr.write('Please supply only one argument\n');
    exit(1);
  }
  var word = args[0];
  generatePermutations(word.split(''));
}

void generatePermutations(List<String> word, [int index]) {
  if (index == 0) {
    print(word.join(''));
  } else {
    if (index == null) {
      index = word.length - 1;
    }
    for (var i = 0; i < index; i++) {
      generatePermutations(word, index - 1);
      var swap = (list, left, right) {
        var saved = list[left];
        list[left] = list[right];
        list[right] = saved;
      };
      swap(word, index % 2 == 0 ? 0 : i, index);
    }
    generatePermutations(word, index - 1);
  }
}
